File: netlify/functions/compress_pdf.js
// netlify/functions/compress_pdf.js
const { PDFDocument } = require('pdf-lib');

exports.handler = async function(event, context) {
    // Handle CORS preflight request (OPTIONS method)
    if (event.httpMethod === 'OPTIONS') {
        return {
            statusCode: 204, // No content needed for preflight success
            headers: {
                'Access-Control-Allow-Origin': '*', // Allow requests from any origin
                'Access-Control-Allow-Methods': 'POST, OPTIONS', // Allow POST and OPTIONS methods
                'Access-Control-Allow-Headers': 'Content-Type', // Allow Content-Type header
            },
            body: ''
        };
    }

    if (event.httpMethod !== 'POST') {
        return {
            statusCode: 405,
            headers: { 'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*' },
            body: JSON.stringify({ error: 'Method Not Allowed. Only POST requests are accepted.' })
        };
    }

    try {
        if (!event.body) {
            return {
                statusCode: 400,
                headers: { 'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*' },
                body: JSON.stringify({ error: 'No request body provided.' })
            };
        }

        // Netlify functions receive the request body as a string, possibly base64 encoded.
        // We expect a JSON payload with pdfFileBase64, compressionLevel, and fileName.
        const requestData = JSON.parse(event.body);
        const pdfFileBase64 = requestData.pdfFileBase64;
        // const compressionLevel = requestData.compressionLevel || 'recommended'; // pdf-lib doesn't use this directly for lossy compression
        const originalFilename = requestData.fileName || 'document.pdf';

        if (!pdfFileBase64) {
            return {
                statusCode: 400,
                headers: { 'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*' },
                body: JSON.stringify({ error: 'No PDF file data provided in the request body.' })
            };
        }

        // Convert base64 PDF data back to a Buffer
        const pdfBytes = Buffer.from(pdfFileBase64, 'base64');
        const originalPdfDoc = await PDFDocument.load(pdfBytes);

        // Create a new PDF document to copy pages into.
        // This process by pdf-lib automatically performs stream compression and object optimization.
        const newPdfDoc = await PDFDocument.create();
        const copiedPages = await newPdfDoc.copyPages(originalPdfDoc, originalPdfDoc.getPageIndices());
        copiedPages.forEach(page => newPdfDoc.addPage(page));

        // Save the new PDF. pdf-lib applies its default compression settings here.
        const compressedPdfBytes = await newPdfDoc.save();

        // Encode the compressed PDF back to base64 for sending to the client
        const compressedPdfBase64 = Buffer.from(compressedPdfBytes).toString('base64');

        // Determine a suitable filename for the downloaded file
        const nameWithoutExt = originalFilename.split('.').slice(0, -1).join('.');
        const compressedFilename = `${nameWithoutExt}_compressed.pdf`;

        // Return the response as JSON containing the base64 encoded file and metadata
        return {
            statusCode: 200,
            headers: {
                'Content-Type': 'application/json', // Indicate that the response body is JSON
                'Access-Control-Allow-Origin': '*', // Crucial for CORS to allow frontend to access
                'Access-Control-Allow-Methods': 'POST, OPTIONS',
                'Access-Control-Allow-Headers': 'Content-Type',
            },
            body: JSON.stringify({
                body: compressedPdfBase64, // The base64 encoded PDF content
                isBase64Encoded: true,     // Signal that 'body' content is base64 encoded
                fileName: compressedFilename, // The suggested filename for download
                originalSize: pdfBytes.length, // Original size in bytes
                compressedSize: compressedPdfBytes.length // Compressed size in bytes
            })
        };

    } catch (error) {
        console.error('Error in Netlify compress_pdf function:', error);
        return {
            statusCode: 500,
            headers: { 'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*' },
            body: JSON.stringify({ error: `Failed to compress PDF: ${error.message}` })
        };
    }
};

File: netlify/functions/compress_pdf.py
# netlify/functions/compress_pdf.py
import json
import base64
import os
from io import BytesIO
from PyPDF2 import PdfReader, PdfWriter # Using PyPDF2 for serverless compatibility

# This is the entry point for Netlify Functions
def handler(event, context):
    """
    Handles PDF compression requests via Netlify Function.
    Expects a base64 encoded PDF file and compressionLevel in the request body.
    Compresses the PDF using PyPDF2 and returns the compressed file as base64.
    """
    try:
        # Check if the request body is present
        if not event.get('body'):
            return {
                'statusCode': 400,
                'headers': { 'Content-Type': 'application/json' },
                'body': json.dumps({'error': 'No request body provided'})
            }

        body = event.get('body')
        is_base64_encoded = event.get('isBase64Encoded', False)

        if is_base64_encoded:
            body = base64.b64decode(body).decode('utf-8')
        
        request_data = json.loads(body) # Assuming client sends JSON payload
        pdf_file_base64 = request_data.get('pdfFileBase64')
        compression_level = request_data.get('compressionLevel', 'recommended')
        original_filename = request_data.get('fileName', 'document.pdf')


        if not pdf_file_base64:
            return {
                'statusCode': 400,
                'headers': { 'Content-Type': 'application/json' },
                'body': json.dumps({'error': 'No PDF file data provided in the request'})
            }

        pdf_bytes = base64.b64decode(pdf_file_base64)
        original_pdf_stream = BytesIO(pdf_bytes)

        reader = PdfReader(original_pdf_stream)
        writer = PdfWriter()

        # Iterate over pages and add them to the writer
        for page_num in range(len(reader.pages)):
            page = reader.pages[page_num]
            writer.add_page(page)

        # The 'compress_streams=True' argument here is important for PyPDF2 compression.
        # This will Flate-encode all streams that are not already compressed.
        compressed_pdf_stream = BytesIO()
        writer.write(compressed_pdf_stream, compress_streams=True)
        compressed_pdf_stream.seek(0) # Rewind to the beginning

        # Prepare response
        compressed_pdf_base64 = base64.b64encode(compressed_pdf_stream.getvalue()).decode('utf-8')

        # Determine a suitable filename for download
        name, ext = os.path.splitext(original_filename)
        compressed_filename = f"{name}_compressed{ext}"

        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json', # Send JSON response
                'Access-Control-Allow-Origin': '*', # Necessary for CORS
                'Access-Control-Allow-Methods': 'POST, OPTIONS',
                'Access-Control-Allow-Headers': 'Content-Type'
            },
            'body': json.dumps({ # Embed base64 and filename in a JSON object
                'body': compressed_pdf_base64,
                'isBase64Encoded': True,
                'fileName': compressed_filename,
                'originalSize': len(pdf_bytes),
                'compressedSize': len(compressed_pdf_stream.getvalue())
            })
        }

    except Exception as e:
        print(f"Error in Netlify function: {e}") # Log to Netlify logs
        return {
            'statusCode': 500,
            'headers': { 'Content-Type': 'application/json' },
            'body': json.dumps({'error': f'Failed to compress PDF: {str(e)}'})
        }

